titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/DefaultContainerResultCodeResolver.java

```diff
@@ -0,0 +1,76 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.master.mesos.kubeapiserver;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.netflix.titus.api.jobmanager.model.job.TaskState;
+import com.netflix.titus.common.util.RegExpExt;
+import com.netflix.titus.common.util.StringExt;
+import com.netflix.titus.master.mesos.MesosConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class DefaultContainerResultCodeResolver implements ContainerResultCodeResolver {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultContainerResultCodeResolver.class);
+
+    private final Function<String, Matcher> invalidRequestMessageMatcherFactory;
+    private final Function<String, Matcher> crashedMessageMatcherFactory;
+    private final Function<String, Matcher> transientSystemErrorMessageMatcherFactory;
+    private final Function<String, Matcher> localSystemErrorMessageMatcherFactory;
+    private final Function<String, Matcher> unknownSystemErrorMessageMatcherFactory;
+
+    @Inject
+    public DefaultContainerResultCodeResolver(MesosConfiguration mesosConfiguration) {
+        this.invalidRequestMessageMatcherFactory = RegExpExt.dynamicMatcher(mesosConfiguration::getInvalidRequestMessagePattern, "invalidRequestMessagePattern", Pattern.DOTALL, logger);
+        this.crashedMessageMatcherFactory = RegExpExt.dynamicMatcher(mesosConfiguration::getCrashedMessagePattern, "crashedMessagePattern", Pattern.DOTALL, logger);
+        this.transientSystemErrorMessageMatcherFactory = RegExpExt.dynamicMatcher(mesosConfiguration::getTransientSystemErrorMessagePattern, "transientSystemErrorMessagePattern", Pattern.DOTALL, logger);
+        this.localSystemErrorMessageMatcherFactory = RegExpExt.dynamicMatcher(mesosConfiguration::getLocalSystemErrorMessagePattern, "localSystemErrorMessagePattern", Pattern.DOTALL, logger);
+        this.unknownSystemErrorMessageMatcherFactory = RegExpExt.dynamicMatcher(mesosConfiguration::getUnknownSystemErrorMessagePattern, "unknownSystemErrorMessagePattern", Pattern.DOTALL, logger);
+    }
+
+    public Optional<String> resolve(TaskState taskState, String reasonMessage) {
+        if (taskState != com.netflix.titus.api.jobmanager.model.job.TaskState.Finished || StringExt.isEmpty(reasonMessage)) {
+            return Optional.empty();
+        }
+
+        if (invalidRequestMessageMatcherFactory.apply(reasonMessage).matches()) {
```

Comment:
minor: add `(matcher, reasonString)` pairs to a collection and `findFirst()` on it to reduce duplication here

Reply:
I would prefer to keep it as it is, in case in the future we need more complex logic when parsing reason messages or checking additional information in the task state.